### PR TITLE
Resolve overflow by limiting influence gain

### PR DESCRIPTION
**Pull request by j-mao**
*This proposal is still a draft and subject to extensive discussion among Teh Devs.*

Exponentiation enables units to grow ridiculously large, sometimes overflowing `Integer.MAX_VALUE`. This is not resolvable by using 64-bit arithmetic, because the rate of exponentiation will easily also exceed `Long.MAX_VALUE`. To resolve this, we propose the following change:

- Limit maximum robot influence to 10^8. Any influence in excess of this will be lost, permanently. This includes influence from all sources:
    - Passive influence gain.
    - Politician empowering (plus muckraker buff).
    - Slanderer influence gain.

Internally we then use 64-bit arithmetic to complete the calculations, and cast back to 32-bits after applying the influence cap.

### CODE REVIEW

File: engine/src/main/battlecode/world/InternalRobot.java [hunk 1]
```diff
@@ -360,25 +372,26 @@ public void empower(int radiusSquared) {
         if (numBots == 0)
             return;
         
-        int convictionToGive = (int) (this.conviction * this.gameWorld.getTeamInfo().getBuff(this.team));
+        long convictionToGive = (long) (this.conviction * this.gameWorld.getTeamInfo().getBuff(this.team));
```
Comment (n8kim1) [inline engine/src/main/battlecode/world/InternalRobot.java:375]: doesn't this compute product as an int, then cast that int to long?
think ((long) this.conviction) * ((long) this.gameWorld...) is better

File: engine/src/main/battlecode/world/InternalRobot.java [hunk 2]
```diff
@@ -360,25 +372,26 @@ public void empower(int radiusSquared) {
...
             if (bot.equals(this))
                 continue;
-            int conv = convictionPerBot;
+            long conv = convictionPerBot;
             if (numBotsWithExtraConviction > 0) {
                 conv++;
                 numBotsWithExtraConviction--;
             }
-            bot.empowered(this, conv, this.team);
+            conv = Math.min(conv, bot.getInfluenceCap() * 2);
```
Comment (n8kim1) [inline engine/src/main/battlecode/world/InternalRobot.java:393]: why x2?

File: engine/src/main/battlecode/world/InternalRobot.java (hunk 1 above)
Comment (n8kim1) [reply engine/src/main/battlecode/world/InternalRobot.java:375]: I could be wrong, maybe casting goes the other way

Comment (j-mao) [reply engine/src/main/battlecode/world/InternalRobot.java:375]: I think `getBuff` returns a double so this computation upcasts to `double`, and then downcasts to `long`.

Comment (j-mao) [reply engine/src/main/battlecode/world/InternalRobot.java:375]: can cast first term to long to be safe

Comment (n8kim1) [reply engine/src/main/battlecode/world/InternalRobot.java:375]: https://repl.it/@n8kim1/EvenHarmfulBoards#Main.java

**Comment by n8kim1** [comment]
humph

**Review by n8kim1 — APPROVED** [review]

**Pull request merged**